File: whiteboard/app/src/main/java/com/ditto/whiteboard/util/SuspendResult.kt
package com.ditto.whiteboard.util

import kotlin.coroutines.cancellation.CancellationException

/** Captures ordinary synchronous failures while preserving cancellation and fatal JVM errors. */
internal inline fun <T> runCatchingException(block: () -> T): Result<T> = try {
  Result.success(block())
} catch (cancelled: CancellationException) {
  throw cancelled
} catch (error: Exception) {
  Result.failure(error)
}

/** Captures ordinary suspend failures while preserving cancellation and fatal JVM errors. */
internal suspend fun <T> runSuspendCatchingPreservingCancellation(
  block: suspend () -> T,
): Result<T> = try {
  Result.success(block())
} catch (cancelled: CancellationException) {
  throw cancelled
} catch (error: Exception) {
  Result.failure(error)
}

File: whiteboard-apple/Packages/WhiteboardCore/Sources/WhiteboardCore/Util/SuspendResult.swift
import Foundation

// Port of util/SuspendResult.kt: capture ordinary failures while rethrowing cancellation.
func runCatchingPreservingCancellation<T>(
  _ block: @Sendable () throws -> T
) throws(CancellationError) -> Result<T, any Error> {
  do {
    return .success(try block())
  } catch is CancellationError {
    throw CancellationError()
  } catch {
    return .failure(error)
  }
}

func runAsyncCatchingPreservingCancellation<T>(
  _ block: @Sendable () async throws -> T
) async throws(CancellationError) -> Result<T, any Error> {
  do {
    return .success(try await block())
  } catch is CancellationError {
    throw CancellationError()
  } catch {
    return .failure(error)
  }
}
